### PR TITLE
Adding feature to allow running difference tests on the same browser 

### PR DESCRIPTION
**Pull request by gitnikos**
This works for one or more threads and has been tested on firefox and chrome.

### CODE REVIEW

File: src/main/java/com/frameworkium/core/ui/tests/BaseTest.java [hunk 1]
```diff
@@ -103,14 +141,28 @@ public static void configureBrowserBeforeTest(String testName) {
...
         } catch (Exception e) {
             baseLogger.warn("Session quit unexpectedly.", e);
         }
     }
 
     /** Shuts down the {@link ExecutorService}. */
     @AfterSuite(alwaysRun = true)
+    public static void tearDownSuite() {
+        if (Property.REUSE_BROWSER.isSpecified()) {
+            for (Driver driver : driverPool) {
```
Comment (ham1) [inline src/main/java/com/frameworkium/core/ui/tests/BaseTest.java]: I think `driverPool.forEach(d-> d.getDriver().quit());` would look neater :)

File: src/main/java/com/frameworkium/core/ui/tests/BaseTest.java [hunk 2]
```diff
@@ -49,6 +49,40 @@
...
+     * Method which returns the next available driver from the pool.
+     * The pool should always have at least one, as it has been initialised with one driver
+     * per thread and each thread pushed its driver back to the pool, when it completes a test.
+     * <p>In the unlikely scenario that the previous execution didn't add the driver back to the pool
+     * for whatever reason, then create a new driver to return.</p>
+     * @return Driver the next available driver from the pool of drivers
+     */
+    public static Driver getNextAvailableDriverFromPool() {
+        if (driverPool.isEmpty()) {
+            driverPool.add(new DriverSetup().instantiateDriver());
```
Comment (ham1) [inline src/main/java/com/frameworkium/core/ui/tests/BaseTest.java]: Would the following also work?

```java
if (driverPool.isEmpty()) {
    return new DriverSetup().instantiateDriver();
}
return driverPool.remove();
```

It feels safer to limit the number of interactions with `driverPool`.

File: .travis.yml [hunk 3]
```diff
@@ -39,7 +39,10 @@ script:
   # Run on Firefox off the grid
   # Currently (Aug-2017) only seems to work with 1 thread on travis
   - mvn verify -B -Dthreads=1 -Dbrowser=firefox -Dwebdriver.gecko.driver=/tmp/geckodriver 2> /dev/null
-  
+
+  # Run on Firefox off the grid by reusing the same browser
+  - mvn verify -B -Dthreads=1 -Dbrowser=firefox -DreuseBrowser=true -Dwebdriver.gecko.driver=/tmp/geckodriver 2> /dev/null
```
Comment (ham1) [inline .travis.yml]: I think this needs testing with at least 2 threads

File: src/main/java/com/frameworkium/core/common/properties/Property.java [hunk 4]
```diff
@@ -39,7 +39,9 @@
     FIREFOX_PROFILE("firefoxProfile"),
     VIDEO_CAPTURE_URL("videoCaptureUrl"),
     CHROME_USER_DATA_DIR("chromeUserDataDir"),
-    CUSTOM_BROWSER_IMPL("customBrowserImpl");
+    CUSTOM_BROWSER_IMPL("customBrowserImpl"),
+    REUSE_BROWSER("reuseBrowser"),
+    THREAD("threads");
```
Comment (ham1) [inline src/main/java/com/frameworkium/core/common/properties/Property.java]: I can't remember if there was a reason why we didn't have this before...
Do tests without `reuseBrowser` still work with multiple threads?
I also think this should probably be called `THREADS` or `THREAD_COUNT` and we can also update `AllureProperties.java` to use it.

File: src/main/java/com/frameworkium/core/ui/driver/AbstractDriver.java [hunk 5]
```diff
@@ -31,7 +31,11 @@
      */
     @Override
     public void tearDown() {
-        this.webDriverWrapper.quit();
+        if (Property.REUSE_BROWSER.isSpecified()) {
```
Comment (ham1) [inline src/main/java/com/frameworkium/core/ui/driver/AbstractDriver.java]: Are we assuming only browsers (i.e. not Appium etc.) here, if so maybe we should have some validation that the `-Dbrowser` is set to something which does indeed behave like a browser? Not sure if this is actually a problem but worth trying out if you (or anyone) has any Appium tests or something which isn't a web browser?

File: .travis.yml (hunk 3 above)
Comment (gitnikos) [reply .travis.yml]: I agree and have been testing locally with 3 threads. However, the comment right above reads "# Currently (Aug-2017) only seems to work with 1 thread on travis"

File: src/main/java/com/frameworkium/core/common/properties/Property.java (hunk 4 above)
Comment (gitnikos) [reply src/main/java/com/frameworkium/core/common/properties/Property.java]: I'll rename to THREADS. The AllureProperties is probably a different ticket.

File: src/main/java/com/frameworkium/core/ui/tests/BaseTest.java (hunk 2 above)
Comment (gitnikos) [reply src/main/java/com/frameworkium/core/ui/tests/BaseTest.java]: Done.

File: src/main/java/com/frameworkium/core/ui/tests/BaseTest.java (hunk 1 above)
Comment (gitnikos) [reply src/main/java/com/frameworkium/core/ui/tests/BaseTest.java]: Done.

File: src/main/java/com/frameworkium/core/ui/tests/BaseTest.java [hunk 6]
```diff
@@ -60,7 +94,11 @@
      */
     @BeforeMethod(alwaysRun = true)
     public static void instantiateDriverObject() {
-        driver.set(new DriverSetup().instantiateDriver());
+        if (!Property.REUSE_BROWSER.isSpecified()) {
```
[the remark this answers is not in the document]
Comment (gitnikos) [reply src/main/java/com/frameworkium/core/ui/tests/BaseTest.java]: Done.

File: src/main/java/com/frameworkium/core/ui/driver/AbstractDriver.java [hunk 7]
```diff
@@ -31,7 +31,11 @@
...
+            webDriverWrapper.manage().deleteAllCookies();
```
[the remark this answers is not in the document]
Comment (gitnikos) [reply src/main/java/com/frameworkium/core/ui/driver/AbstractDriver.java]: No. For clearing caches, see last comment [here](https://github.com/seleniumhq/selenium-google-code-issue-archive/issues/40).

There might be some workarounds with using a Firefox profile that has caching disabled, but this kinda defeats the purpose of speeding up execution via browser reuse. 

I think we could still offer the browser-reuse feature even without any cache clearing.

[the remark this answers is not in the document]
Comment (ham1) [reply src/main/java/com/frameworkium/core/ui/driver/AbstractDriver.java]: Ok, I forgot about this, probably contributed to why we changed it to set up a new browser each test. I think this is fine as long as we document this limitation.

**Pull request merged**